### DISCUSSION TITLE
Is there a way to add custom user-specific settings when building containers with Dev Container?

### DISCUSSION THREAD

**Discussion by bells17** [devcontainers/community#134] [Q&A]
Hello, I have a question regarding the use of Dev Containers for building containers. Is there a way to incorporate each user's custom settings? If so, could you please explain how?

From what I've checked from the website (https://containers.dev/), it appears that the `devcontainer.json` file does not offer any methods for adding user-specific settings and commands, for instance, using `onCreateCommand` or other extension points.

In a development environment, there are necessary settings and tools that all users need, and others that are not universally required. For example, in the `kubernetes/kubernetes` project, the method of running e2e tests seems to vary among users.
https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#e2e-tests
https://github.com/kubernetes-sigs/kind/issues/1181#issuecomment-567001724

I am currently working on a PR to introduce dev containers into `kubernetes/kubernetes`, and it is a challenging issue to decide how many such tools should be installed.
https://github.com/kubernetes/kubernetes/pull/124584

Not limited to `kubernetes/kubernetes`, there are often cases where having tools like `jq` installed personally can make development more convenient, even if they might not be significant enough to include in a repository's `devcontainer.json`. For instance, installing such tools individually after building the Dev Container can be inconvenient because if the Dev Container is rebuilt, you must reinstall the tools.

If there was a feature in Dev Containers that allowed each user to automatically add their custom settings outside of devcontainer.json during container creation, it could significantly enhance the flexibility and potentially make Dev Containers more widely adopted by developers. I am curious if there are already established best practices or solutions for this issue. If you have any experiences or knowledge to share, I would greatly appreciate your input.

**Comment by samruddhikhandale** [comment]
Hi 👋 

Thanks for a detailed description and yes, there is a way to incorporate each user's custom settings in dev containers. You can use a combination of `devcontainer.json` and `.env` files to achieve this.

The `devcontainer.json` file is used to define the container configuration, while the `.env` file can be used to store user-specific settings. The `.env` file is not typically checked into source control, allowing each developer to have their own settings.

Here's a step-by-step plan:

- Define the necessary settings in the devcontainer.json file that will be common to all users.
- For user-specific settings, use environment variables in the devcontainer.json file. These variables can be defined in the .env file.
- Instruct each user to create their own .env file with their specific settings.
- Use these environment variables in your scripts or application code as needed.

If you use Codespaces, then you can achieve the same with dotfiles. Refer to https://docs.github.com/en/codespaces/setting-your-user-preferences/personalizing-github-codespaces-for-your-account#dotfiles

Side note, looping in @craiglpeters who was already working on a PR to contribute a dev container to `kubernetes/kubernetes`. From the discussions we had, that seemed tricky due to the complex infra for kubernetes. However, it's still possible Good luck!

**Reply by bells17** [reply]
@samruddhikhandale Thank you for your insightful response and the practical suggestions. May I ask a few additional questions?

> The `devcontainer.json` file is used to define the container configuration, while the .env file can be used to store user-specific settings.
> The `.env` file is not typically checked into source control, allowing each developer to have their own settings.

Could you share some examples of repositories that have implemented this approach? I would like to refer to practical examples of the configuration method.

> Use these environment variables in your scripts or application code as needed.

Regarding this point, in the case where instead of installing pre-defined optional tools, users prepare and execute their own scripts, would the following approach be appropriate?

- Configure `.gitignore` to exclude the `.env` file from the Git repository, ensuring user-specific settings remain confidential.
- Designate a specific directory (e.g., `.devcontainer/user-scripts`) as a location for users to place their dev container scripts, and configure `.gitignore` to exclude this directory.
- Users place their scripts in this directory (e.g., `.devcontainer/user-scripts/onCreateContainer.sh`).
- In the .env file, users set the `ON_CREATE_CONTAINER_COMMAND` environment variable to point to their script (e.g., `.devcontainer/user-scripts/onCreateContainer.sh`).
- In `devcontainer.json`, the `onCreateContainerCommand` is configured to execute the script specified by the `ON_CREATE_CONTAINER_COMMAND` environment variable, if it is set.

> Side note, looping in @craiglpeters who was already working on a PR to contribute a dev container to `kubernetes/kubernetes`. From the discussions we had, that seemed tricky due to the complex infra for kubernetes. However, it's still possible Good luck!

Yes, @craiglpeters was working on introducing a dev container to `kubernetes/kubernetes`, but he mentioned having difficulty allocating time for this task in our conversation (https://kubernetes.slack.com/archives/C1TU9EB9S/p1714223550391479?thread_ts=1713701062.612009&cid=C1TU9EB9S). After confirming with him that I could proceed, I have created a PR for this.

I have requested a review from @craiglpeters on my PR. If they find the time, their comments would be greatly appreciated.

**Reply by bells17** [reply]
My apologies, please allow me to ask additional questions.

> Thanks for a detailed description and yes, there is a way to incorporate each user's custom settings in dev containers. You can use a combination of devcontainer.json and .env files to achieve this.  

Could you please direct me to the section on the official website (https://containers.dev/) that discusses this method? Additionally, if this method is not currently documented, are there any plans to include such information on the official website in the future?

**Reply by bells17** [reply]
@samruddhikhandale, I'd appreciate your thoughts on my previous post when you have a moment. Please let me know if you need any clarification or additional information.
